core/client/src/main/java/alluxio/client/file/policy/DeterministicHashPolicy.java

```diff
@@ -0,0 +1,136 @@
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.client.file.policy;
+
+import alluxio.client.block.BlockWorkerInfo;
+import alluxio.master.block.BlockId;
+import alluxio.wire.WorkerNetAddress;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Lists;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Random;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * This policy maps blockId to several deterministic Alluxio workers. The number of workers a block
+ * can be mapped to can be passed through the constructor. The default is 1. It skips the workers
+ * that doesn not have enough capacity to hold the block.
+ */
+@NotThreadSafe
+public final class DeterministicHashPolicy implements BlockLocationPolicy {
+  private final Random mRandom = new Random();
+  private List<BlockWorkerInfo> mWorkerInfoList;
+  private int mOffset;
+  private boolean mInitialized = false;
+  private int mShards = 1;
+
+  /**
+   * Constructs a new {@link DeterministicHashPolicy}.
+   */
+  public DeterministicHashPolicy() {}
+
+  /**
+   * Constructs a new {@link DeterministicHashPolicy}.
+   *
+   * @param numShards the number of shards a block's traffic can be sharded to
+   */
+  public DeterministicHashPolicy(Integer numShards) {
+    Preconditions.checkArgument(numShards >= 1);
+    mShards = numShards;
+  }
+
+  @Override
+  public WorkerNetAddress getWorkerForBlock(Iterable<BlockWorkerInfo> workerInfoList,
+      long blockId, long blockSizeBytes) {
+    if (!mInitialized) {
+      mWorkerInfoList = Lists.newArrayList(workerInfoList);
+      Collections.sort(mWorkerInfoList, new Comparator<BlockWorkerInfo>() {
+        @Override
+        public int compare(BlockWorkerInfo o1, BlockWorkerInfo o2) {
+          return o1.getNetAddress().toString().compareToIgnoreCase(o2.getNetAddress().toString());
+        }
+      });
+      mOffset = (int) (BlockId.getContainerId(blockId) % (long) mWorkerInfoList.size());
+      mInitialized = true;
+    }
+
+    List<WorkerNetAddress> workers = new ArrayList<>();
+    // Try the next one if the worker mapped from the blockId doesn't work until all the workers
+    // are examined.
+    int index =
+        (int) ((mOffset + BlockId.getSequenceNumber(blockId)) % (long) mWorkerInfoList.size());
+    for (int i = 0; i < mWorkerInfoList.size(); i++) {
+      WorkerNetAddress candidate = mWorkerInfoList.get(index).getNetAddress();
+      BlockWorkerInfo workerInfo = findBlockWorkerInfo(workerInfoList, candidate);
```

Comment:
hmm, why O(num_workers * num_shards)? It is O(num_shards) when the block can fit all the workers.  The worst case is O(num_workers)?